### ISSUE TITLE
Act: Build fails with `sh: 1: cargo: not found`

### ISSUE THREAD

**Issue by Sellig6792** [cross-rs/cross#1321]
Hello, I've just finished the new version of my project. But when I did the PR an error occurred in the workflow. So I decided to revert the merge and tested the workflow locally with Act. Unfortunately, I discovered that, locally, I had one more error.
For Information MacOS isn't running with Act.

The cross version is the `v0.2.5`

The Error
```
| [command]cross build --release --target x86_64-pc-windows-gnu
| info: downloading component 'rust-src'
| info: installing component 'rust-src'
| sh: 1: cargo: not found
```
The Job:
```yml
deploy_building:
  name: Deploy Building
  needs: [ test ]
  strategy:
    matrix:
      os: [ ubuntu-latest ] # Default os for build
      target: [ aarch64-unknown-linux-gnu,
                i686-pc-windows-gnu, x86_64-pc-windows-gnu,
                i686-unknown-linux-gnu, x86_64-unknown-linux-gnu,
                x86_64-apple-darwin ]
      exclude: # Do not build for macOS on Linux
        - target: x86_64-apple-darwin
          os: ubuntu-latest
      include:
        - target: x86_64-apple-darwin
          os: macos-latest
  # Runs on latest ubuntu by default except for windows targets
  runs-on: ${{ matrix.os }}
  steps:
    - name: Checkout
      uses: actions/checkout@v1
    - name: Install rust
      uses: actions-rs/toolchain@v1
      with:
        toolchain: stable
        profile: minimal
        override: true
        target: ${{ matrix.target }}
    - name: Build release
      uses: actions-rs/cargo@v1
      with:
        use-cross: true
        command: build
        args: --release --target ${{ matrix.target }}
```

[Here ](https://github.com/cross-rs/cross/files/12454570/act_log_verbose_one_target.txt)  is the verbose log for only one target (windows x86_64)

Thanks to all the people who will try to solve my problem

**Comment by Emilgardis** [comment]
you'll probably need `CROSS_CONTAINER_IN_CONTAINER=true` for `act` , `act` uses docker to containerize the actions and cross can't properly mount in those cases.

Pass in `-v` in the command to cross to see what it's doing.

Also, actions-rs is unmaintained, I'd highly recommend you find an alternative

**Comment by Emilgardis** [comment]
I don't know how act works and I've never gotten it to work myself.

I suspect it's setting something weirdly, I'm not sure how to solve it
Is it possible to use docker from within act?

The error comes from that it's not possible to re-mount inside a running container without access to the underlying filesystem.

Another possible solution would be to use `CROSS_REMOTE`, see https://github.com/cross-rs/cross/wiki/FAQ#gitlab-ci

**Comment by Sellig6792** [comment]
I didn't find anything about how to add `-v` to docker through act. Do you know how to do that ?
Maybe I should open an issue on the repository of Act or edit and build Act myself...

**Comment by Maximilian-Staab** [comment]
Pretty sure this is a docker-in-docker-in-docker issue, at least it was in my case. 

Basically cross tries to mount volumes into the the build container, but because cross/action-rs itself is running in a container, the mounts are relative to the act runner, not the container cross is called from.

I had to mount and use a volume like this:
```
jobs:
  your-job:
    container:
      volumes: # these have to be at this strange path, because the cross docker command will mount these volumes
        - /opt/repo:/opt/repo # can't be anonymous, because the volume will be deanonymized by cross

    steps:
      - name: Checkout repository
        uses: actions/checkout@v3
      - name: Copy repository for cross
        run: cp -a .. /opt/repo
      - name: Install cross-rs
        env: 
          CROSS_CONTAINER_IN_CONTAINER: true
        run: cargo install cross

      - name: Build Package
        working-directory: /opt/repo/snapper-rs
        env: 
          CROSS_CONTAINER_IN_CONTAINER: true
        run: cross build --release --target armv7-unknown-linux-gnueabihf -v
 ```
This is just an example, you need to clone your repo and install the rust toolchain as well. 

If you use a `Dockerfile` for your cross build container, you need to install the rust toolchain in there manually. Usually the mounts would take care of that.

**Comment by Sellig6792** [comment]
> Pretty sure this is a docker-in-docker-in-docker issue, at least it was in my case.
> 
> Basically cross tries to mount volumes into the the build container, but because cross/action-rs itself is running in a container, the mounts are relative to the act runner, not the container cross is called from.
> 
> I had to mount and use a volume like this:
> 
> ```
> jobs:
>   your-job:
>     container:
>       volumes: # these have to be at this strange path, because the cross docker command will mount these volumes
>         - /opt/repo:/opt/repo # can't be anonymous, because the volume will be deanonymized by cross
> 
>     steps:
>       - name: Checkout repository
>         uses: actions/checkout@v3
>       - name: Copy repository for cross
>         run: cp -a .. /opt/repo
>       - name: Install cross-rs
>         env: 
>           CROSS_CONTAINER_IN_CONTAINER: true
>         run: cargo install cross
> 
>       - name: Build Package
>         working-directory: /opt/repo/snapper-rs
>         env: 
>           CROSS_CONTAINER_IN_CONTAINER: true
>         run: cross build --release --target armv7-unknown-linux-gnueabihf -v
> ```
> 
> This is just an example, you need to install the rust toolchain as well.
> 
> If you use a `Dockerfile` for your cross build container, you need to install the rust toolchain in there manually. Usually the mounts would take care of that.

Could you adapt it to my workflow ?
Thanks

**Comment by Sitin** [comment]
I've created setup for `CROSS_REMOTE` similar to what GitLab uses by using Docker-in-Docker official image. It seems that `cross` in remote mode is messing with container layers and breaks containers.

Results can be found in the [`gist`](https://gist.github.com/Sitin/db0de5bdb98d648b320f5a77137b8ece).

You also can check the [repo](https://gitlab.com/ziatin/cross-dind/-/tree/1b7ae70a6981e9de2732f01101623a1817864ab3) with code I've created for my experiments.
